gprofiler/merge.py

```diff
@@ -94,7 +100,11 @@ def merge_perfs(perf_all: Iterable[Mapping[str, str]], process_perfs: Mapping[in
         if process_perf_count > 0:
             ratio = perf_all_count / process_perf_count
             for stack, count in process_stacks.items():
-                full_stack = ";".join([process_names[pid], stack])
-                new_samples[full_stack] += round(count * ratio)
-
-    return "\n".join((f"{stack} {count}" for stack, count in new_samples.items()))
+                stack_line = ";".join([process_names[pid], stack])
+                stack_line_with_container_name = docker_client.get_stack_line_with_container_name(stack_line, pid)
+                new_samples[stack_line_with_container_name] += round(count * ratio)
+    container_names = docker_client.get_container_names()
+    docker_client.reset_cache()
```

Comment:
please include only `get_container_names` that currently have stack sampled from within them - not all in the cache / running now
UNLESS
are you resetting the cache every time? @Jongy is this legit performance wise or should we prolong the values across runs?

Reply:
> please include only get_container_names that currently have stack sampled from within them - not all in the cache / running now

We attempt to get the container ID only for PIDs present in the stackcollapsed output, so that should be alright.

> are you resetting the cache every time? @Jongy is this legit performance wise or should we prolong the values across runs?

Resetting the PIDs cache is okay (and desired). Reading `/proc/pid/cgroup` per PID should be okay.

Resetting the container ID <-> name cache should be avoided (like I mentioned earlier). We need to measure the impact, but I presume that querying the Docker daemon for tens/hundreds of container names every interval is *not* desired.

Reply:
Actually, @alonbentamar I have a better suggestion: you will manage an ID <-> name cache, but instead of querying Docker for every missing mapping, you will query just *once* via `_client.containers.list(...)` and update the cache with all results. This way we avoid multiple Docker API queries.

Reply:
Good now.